x-pack/legacy/plugins/ml/public/util/custom_url_utils.ts

```diff
@@ -0,0 +1,206 @@
+) {
+  if (isKibanaUrl(urlConfig) === true) {
+    return buildKibanaUrl(urlConfig, record);
+  } else {
+    const urlPath = replaceStringTokens(urlConfig.url_value, record, false);
+    return urlPath;
+  }
+}
+
+// Opens the specified URL in a new window. The behaviour (for example whether
+// it opens in a new tab or window) is determined from the original configuration
+// object which indicates whether it is opening a Kibana page running on the same server.
+// fullUrl is the complete URL, including the base path, with any dollar delimited tokens
+// from the urlConfig having been substituted with values from an anomaly record.
+export function openCustomUrlWindow(fullUrl: string, urlConfig: UrlConfig) {
+  // Run through a regex to test whether the url_value starts with a protocol scheme.
+  if (/^(?:[a-z]+:)?\/\//i.test(urlConfig.url_value) === false) {
+    window.open(fullUrl, '_blank');
+  } else {
+    // Add noopener and noreferrr properties for external URLs.
+    const newWindow = window.open(fullUrl, '_blank', 'noopener,noreferrer');
+
+    // Expect newWindow to be null, but just in case if not, reset the opener link.
+    if (newWindow !== undefined && newWindow !== null) {
+      newWindow.opener = null;
+    }
+  }
+}
+
+// Returns whether the url_value of the supplied config is for
+// a Kibana Discover or Dashboard page running on the same server as this ML plugin.
+function isKibanaUrl(urlConfig: UrlConfig) {
+  const urlValue = urlConfig.url_value;
+  return urlValue.startsWith('kibana#/discover') || urlValue.startsWith('kibana#/dashboard');
+}
+
+/**
+ * Escape any double quotes in the value for correct use in KQL.
+ */
+function escapeForKQL(value: string): string {
+  return value.replace(/\"/g, '\\"');
+}
+
+// Builds a Kibana dashboard or Discover URL from the supplied config, with any
+// dollar delimited tokens substituted from the supplied anomaly record.
+function buildKibanaUrl(urlConfig: UrlConfig, record: AnomalyRecordSource) {
+  const urlValue = urlConfig.url_value;
+  const URL_LENGTH_LIMIT = 2000;
+
+  const isLuceneQuery = urlValue.includes('language:lucene');
+  const queryLanguageEscapeCallback = isLuceneQuery ? escapeForElasticsearchQuery : escapeForKQL;
+
+  type GetResultTokenValue = (v: string) => string;
+
+  // Compose callback for characters escaping and encoding.
+  const getResultTokenValue: GetResultTokenValue = flow(
+    queryLanguageEscapeCallback,
+    // Kibana URLs used rison encoding, so escape with ! any ! or ' characters
+    (v: string): string => v.replace(/[!']/g, '!$&'),
+    encodeURIComponent
+  );
+
+  return String(urlValue)
+    .replace('$earliest$', record.earliest)
+    .replace('$latest$', record.latest)
+    .replace(
+      /(.+query:')([^']*)('.+)/,
+      (fullMatch, prefix: string, queryString: string, postfix: string) => {
+        const [resultPrefix, resultPostfix] = [prefix, postfix].map(str =>
+          str.replace(/\$(\w+)\$/g, (match, name: string) => {
```

Comment:
this replace function is the same as the one on line 179, could they be combined?
a good function name might avoid the need for comments explaining what it's doing